src/hotspot/share/runtime/threads.cpp

```diff
@@ -564,7 +564,12 @@ jint Threads::create_vm(JavaVMInitArgs* args, bool* canTryAgain) {
   status = init_globals2();
   if (status != JNI_OK) {
     Threads::remove(main_thread, false);
-    main_thread->smr_delete();
+    // It is possible that we managed to fully initialize Universe but have then
+    // failed by throwing an exception. In that case our caller JNI_CreateJavaVM
+    // will want to report it, so we can't delete the main thread.
+    if (!main_thread->has_pending_exception()) {
+      main_thread->smr_delete();
+    }
```

Comment:
Also init_globals2 should have a better name, like init_objects_and_globals() or something like that.  But that's not this issue.

Reply:
I was later wondering the same thing. Here's my rationale for putting it here
instead of in the JNI code (i.e. agreeing with Coleen's suggestion).

I think neither place is really correct. Failure to create and initialize the
VM ought to always propogate out to the JNI client to decide what to do about
the failure. But we don't do that, not only here but all over the place. These
(and many other) failures are treated as grounds for terminating the process
containing the VM, regardless of what else that process might be doing. This
limits the utility of the VM in ways similar to not being able to have
multiple VMs (at the same time or serially) in one process. Oh well.

Given that, putting the vm_exit_during_initialization() in create_vm seems
simpler.  It's not even clear create_vm should ever return failure with
canTryAgain false; maybe it should always terminate if there's a non-retryable
failure.  But perhaps that's for a later day.

Reply:
I agree that init_globals2 isn't a great name. I failed to come up with
something better, either now or when I added it.  Anyone with something better
should go for it.

Reply:
Sorry but this is getting a bit out of hand. Rearranging the deckchairs really doesn't change anything. You could argue that everything should happen in create_vm so JNI_CreateJavaVM is just a minimal wrapper, but that is not what fixing this issue is about. All I'm trying to do here is avoid the crash when there is no longer a current thread.

Reply:
New version committed.

Reply:
New version reverted. You can't just do the exception check here because there is other code below that can thrown exceptions.